src/HashMap.ts

```diff
@@ -610,15 +612,34 @@ export class HashMap<K,V> implements IMap<K,V> {
         return true;
     }
 
+    private mapHash(): number {
+        return this.hamt.fold(
+            (acc: number, value: V, key: K & WithEquality) =>
+                acc + (getHashCode(key) ^ getHashCode(value)), 0);
+    }
+
     /**
      * Get a number for that object. Two different values
      * may get the same number, but one value must always get
      * the same number. The formula can impact performance.
      */
     hashCode(): number {
-        return this.hamt.fold(
-            (acc: number, value: V, key: K & WithEquality) =>
-                acc + fieldsHashCode(key, value), 0);
+        // references:
+        // https://github.com/clojure/clojure/blob/5ffe3833508495ca7c635d47ad7a1c8b820eab76/src/jvm/clojure/lang/APersistentMap.java#L98
+        // https://github.com/AdoptOpenJDK/openjdk-jdk11/blob/19fb8f93c59dfd791f62d41f332db9e306bc1422/src/java.base/share/classes/java/util/HashMap.java#L296
+
+        // Both the Clojure and AdoptOpenJDK references calculate the bitwise XOR
+        // of the key/value pairs. Because the Clojure implementation is immutable,
+        // like ours, we can also cache the result.
```

Comment:
I think people generally won't mutate the values inside of the hashmaps directly and the API (e.g. put) returns a new instance of the HashMap, so I think this change is safe and won't break anything, unless people are doing really weird things.

We use a "DataClass" abstraction instead of vanilla JavaScript objects that use the same mechanism for caching the hash codes. The hashmaps are usually properties of these DataClass objects, so caching the hash code of the hashmaps as well won't significantly impact performance for us.

It is up to you, if you are not sure about it, let's just revert the caching, as it can be achieved using a wrapper like our DataClass, too.

Reply:
In case anyone is interested, this is our data class wrapper:
``````typescript
import { HasEquals, Option, getHashCode } from "prelude-ts";

import isDataField from "../../../utils/isDataField";
import transformDataFieldsToPreludeJson from "../../../utils/transformDataFieldsToPreludeJson";

export type DataClassFields<E extends DataClass> = Omit<
  E,
  keyof HasEquals | "_hashCode" | "toJSON"
>;

const hasHashCode = (value: unknown): value is { hashCode(): number } =>
  // @ts-expect-error value needs to be unknown
  typeof value.hashCode === "function";

export default abstract class DataClass implements HasEquals {
  private _hashCode: Option<number> = Option.none();

  protected constructor(fields: DataClassFields<DataClass>) {
    this.init(fields);
  }

  // Needed because JavaScript's classes suck
  protected init(fields: Record<string, unknown>): void {
    Object.entries(fields).forEach(([key, value]) => {
      if (isDataField([key, value]))
        // @ts-expect-error we need to generically index the DataFields
        this[key] = value;
    });
  }

  public toJSON(): Record<string, unknown> {
    return transformDataFieldsToPreludeJson(this as Record<string, unknown>);
  }

  public equals(other: unknown): boolean {
    return (
      !other ||
      this === other ||
      !hasHashCode(other) ||
      this.hashCode() === other.hashCode()
    );
  }

  public hashCode(): number {
    return this._hashCode.getOrCall(() => {
      const hashCode = Object.entries(this)
        .sort(([ka], [kb]) => ka.localeCompare(kb))
        .reduce(
          (acc, [k, v]) =>
            isDataField([k, v])
              ? // casting to string, to keep the number within max/min bounds
                getHashCode(`${acc},${getHashCode(v)}`)
              : acc,
          0
        );

      this._hashCode = Option.of(hashCode);

      return hashCode;
    });
  }
}
```

Reply:
Note that the implementation is not necessarily perfect (e.g. casting the value to a string before calculating the hash code ;), but it does it's job in our application,

Reply:
i mean at this point this library is really stable. probably best not to change behavior like that, and if we did, then we should probably do it for all the collections (vector, hashset and so on). if you do it for all the collections, i would probably take it and then release as 1.1.0 to make it clear there was some behavior change. so presumably remove that bit yes.

Reply:
I've removed the caching.